### PR TITLE
Disable optimization for constant functions yield unserializable objects

### PR DESCRIPTION
**Pull request by hellium01**
Some constant function can return result types that cannot be translated
back to expression. We will want to disable optimization in planning
stage and leave the RowExpression optimizer to inline the results of
those functions.

Issue #12213

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java [hunk 1]
```diff
@@ -913,7 +914,8 @@ protected Object visitFunctionCall(FunctionCall node, Object context)
             }
 
             // do not optimize non-deterministic functions
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java]: The comment is out of date; may need to update.

**Comment by highker** [comment]
To add up @hellium01's comment, all type conversions are happening in `TypeRegistry` in presto-main that cannot be overridden/inherited. customized type coercion is needed.

**Comment by martint** [comment]
Have you tried making FbJsonPath serializable?

**Comment by highker** [comment]
That is in `LiteralEncoder` in presto-main with the help of magic literal? cc: @wenleix

**Comment by hellium01** [comment]
Yeah, calling it serializable is a little misleading. We need to be able to encode it into literal so that we can rewrite the expression and send it to workers.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java [hunk 2]
```diff
@@ -1044,6 +1046,12 @@ protected Object visitLikePredicate(LikePredicate node, Object context)
                     optimizedEscape);
         }
 
+        private boolean typeCanSerialize(Type type)
+        {
+            return isSupportedLiteralType(type) || type.getJavaType().isAssignableFrom(Block.class)
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java]: Why `Block` is considered as `serializable`? 

Also, my understanding is it's about whether the value can be encoded in plan (and distribute to worker nodes) right? So maybe call it `typeCanBeEncodedInPlan`?

**Comment by wenleix** [comment]
@hellium01 

> Yeah, calling it serializable is a little misleading. We need to be able to encode it into literal so that we can rewrite the expression and send it to workers. Encoding custom type to literal is not possible at the moment.

Yeah, it's the limitation of the current Type plugin system. There is no easy way to add type coercion  rule and magic literal encoding rule for added types -- they are hard coded in `LiteralEncoder` and `TypeRegistry.coerceTypeBase`.

We could ad a comment stating once added type supports customizing coercion and magic literal encoding rules, we could remove that. But for now this looks like the only way.

**Comment by wenleix** [comment]
Also remember to fix the commit message title. "unserializable" is a bit confusing, as discussed earlier.

**Comment by wenleix** [comment]
@hellium01 : Do you still need this PR? Maybe either merge it or close it? Thanks!

**Pull request closed without merging**